pkg/uki/common.go

```diff
@@ -153,3 +161,140 @@ func copyFile(src, dst string) error {
+	info, err := fs.Stat(artifact)
+	if errors.Is(err, os.ErrNotExist) {
+		logger.Warnf("%s does not exist", artifact)
+		return fmt.Errorf("%s does not exist", artifact)
+	} else if errors.Is(err, os.ErrPermission) {
+		logger.Warnf("%s permission denied. Can't read file", artifact)
+		return fmt.Errorf("%s permission denied. Can't read file", artifact)
+	} else if err != nil {
+		return err
+	}
+	if info.Size() == 0 {
+		logger.Warnf("%s file is empty denied", artifact)
+		return fmt.Errorf("%s file has zero size", artifact)
+	}
+	logger.Logger.Debug().Str("what", artifact).Msg("Reading artifact")
+
+	// MMAP the file, seems to save memory rather than reading the full file
+	// Unfortunately we have to do some type conversion to keep using the v1.Fs
+	f, err := fs.Open(artifact)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+	// type conversion, ugh
+	fOS := f.(*os.File)
+	data, err := mmap.Map(fOS, mmap.RDONLY, 0)
+	defer data.Unmap()
+	if err != nil {
+		return err
+	}
+
+	// Get sha256 of the artifact
+	// Note that this is a PEFile, so it's a bit different from a normal file as there are some sections that need to be
+	// excluded when calculating the sha
+	logger.Logger.Debug().Str("what", artifact).Msg("Parsing PE artifact")
+	file, _ := peparser.NewBytes(data, &peparser.Options{Fast: true})
+	err = file.Parse()
+	if err != nil {
+		logger.Logger.Error().Err(err).Msg("parsing PE file for hash")
+		return err
+	}
+
+	logger.Logger.Debug().Str("what", artifact).Msg("Checking if its an EFI file")
+	// Check for proper header in the efi file
+	if file.DOSHeader.Magic != peparser.ImageDOSZMSignature && file.DOSHeader.Magic != peparser.ImageDOSSignature {
+		logger.Error(fmt.Errorf("no pe file header: %d", file.DOSHeader.Magic))
+		return fmt.Errorf("no pe file header: %d", file.DOSHeader.Magic)
+	}
+
+	// Get hash to compare in dbx if we have hashes
+	hashArtifact := hex.EncodeToString(file.Authentihash())
+
+	logger.Logger.Debug().Str("what", artifact).Msg("Getting DB certs")
+	// We need to read the current db database to have the proper certs to check against
+	db, err := efi.Getdb()
+	if err != nil {
+		logger.Logger.Error().Err(err).Msg("Getting DB certs")
+		return err
+	}
+
+	dbCerts := signatures.ExtractCertsFromSignatureDatabase(db)
+
+	logger.Logger.Debug().Str("what", artifact).Msg("Getting signatures from artifact")
+	// Get signatures from the artifact
+	sigs, err := pecoff.GetSignatures(data)
+	if err != nil {
+		return fmt.Errorf("%s: %w", artifact, err)
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("no signatures in the file %s", artifact)
+	}
+
+	logger.Logger.Debug().Str("what", artifact).Msg("Getting DBX certs")
+	dbx, err := efi.Getdbx()
+	if err != nil {
+		logger.Logger.Error().Err(err).Msg("getting DBX certs")
+		return err
+	}
+
+	// First check the dbx database as it has precedence, on match, return immediately
+	for _, k := range *dbx {
```

Comment:
this logic seems to me like it has to be similar to one we already have, can we somehow have a single function for this if that is the case?

Reply:
We got something like this in `dbCerts := signatures.ExtractCertsFromSignatureDatabase(db)` which gets the signatures but for DBX is not exactly similar as it can be of different types (hash, cert or even a third option which I have not encountered) so this is a bit more convoluted that the normal PK/DB/KEK so I would leave it as is as its very specific to this usecase.